suite-native/fiat-rates/src/fiatRatesThunks.ts

```diff
@@ -0,0 +1,201 @@
+    async ({ account, txs, localCurrency }: UpdateTxsFiatRatesThunkPayload, { dispatch }) => {
+        if (txs?.length === 0 || isTestnet(account.symbol)) return;
+
+        const timestamps = txs.map(tx => tx.blockTime!);
+        const response = await TrezorConnect.blockchainGetFiatRatesForTimestamps({
+            coin: account.symbol,
+            timestamps,
+            currencies: [localCurrency],
+        });
+        try {
+            const results = response.success
+                ? response.payload
+                : await getFiatRatesForTimestamps({ symbol: account.symbol }, timestamps);
+
+            if (results && 'tickers' in results) {
+                dispatch(
+                    fiatRatesActions.updateTransactionFiatRate(
+                        txs.map((tx, i) => ({
+                            txid: tx.txid,
+                            updateObject: { rates: results.tickers[i]?.rates },
+                            account,
+                            ts: Date.now(),
+                        })),
+                    ),
+                );
+            }
+        } catch (error) {
+            console.error(error);
+        }
+    },
+);
+
+const fetchFiatRate = async (
+    ticker: TickerId,
+    fiatCurrency: FiatCurrencyCode,
+): Promise<number | undefined | null> => {
+    const { symbol, tokenAddress, mainNetworkSymbol } = ticker;
+    const { success, payload } = await TrezorConnect.blockchainGetCurrentFiatRates({
+        coin: mainNetworkSymbol || symbol,
+        token: tokenAddress,
+        currencies: [fiatCurrency],
+    });
+
+    const rate = success ? payload.rates?.[fiatCurrency] : null;
+    if (rate) return rate;
+
+    return fetchCurrentFiatRates(ticker).then(res => res?.rates?.[fiatCurrency]);
+};
+
+const fetchLastWeekRate = async (
+    ticker: TickerId,
+    fiatCurrency: FiatCurrencyCode,
+): Promise<number | undefined | null> => {
+    const weekAgoTimestamp = getUnixTime(subWeeks(new Date(), 1));
+    const timestamps = [weekAgoTimestamp];
+
+    const { success, payload } = await TrezorConnect.blockchainGetFiatRatesForTimestamps({
+        coin: ticker.symbol,
+        token: ticker.tokenAddress,
+        timestamps,
+        currencies: [fiatCurrency],
+    });
+
+    const rate = success ? payload.tickers?.[0]?.rates?.[fiatCurrency] : null;
+    if (rate) return rate;
+
+    return fetchLastWeekFiatRates(ticker, fiatCurrency).then(
+        res => res?.tickers?.[0]?.rates?.[fiatCurrency],
+    );
+};
+
+const fetchFn: Record<RateType, typeof fetchFiatRate> = {
+    current: fetchFiatRate,
+    lastWeek: fetchLastWeekRate,
+};
+
+type UpdateCurrentFiatRatesThunkPayload = {
+    ticker: TickerId;
+    fiatCurrency: FiatCurrencyCode;
+    rateType?: RateType;
+};
+
+export const updateFiatRatesThunk = createThunk(
+    `${actionPrefix}/updateFiatRates`,
+    async ({ ticker, fiatCurrency, rateType = 'current' }: UpdateCurrentFiatRatesThunkPayload) => {
```

Comment:
Do we need to send `fiatCurrency` as an argument? What about reading it with a `selectFiatCurrency` selector inside of the thunk? Just an idea.

Reply:
Yes, but I want to move this to common ideally in close future and currency selector is specific for desktop/mobile. This is more future-proof.